src/badguy/goldbomb.cpp

```diff
@@ -141,15 +162,113 @@ GoldBomb::active_update(float dt_sec)
     }
     return;
   }
-  if (is_grabbed())
+
+  if (m_frozen) return;
+
+  if ((tstate == STATE_FLEEING || tstate == STATE_CORNERED) && on_ground() && might_fall(FLEEING_MAX_DROP_HEIGHT+1))
+  {
+    // also check for STATE_CORNERED just so
+    // the bomb doesnt automatically turn around
+    cornered();
     return;
+  }
   WalkingBadguy::active_update(dt_sec);
+
+  MovingObject* obj = nullptr;
+  std::vector<MovingObject*> objs = Sector::get().get_nearby_objects(get_bbox().get_middle(), SAFE_DIST);
+  for (size_t i = 0; i < objs.size(); i++)
+  {
+    obj = objs[i];
+
+    auto player = dynamic_cast<Player*>(obj);
+    if (player && !player->get_ghost_mode()) break;
+
+    auto haywire = dynamic_cast<Haywire*>(obj);
+    if (haywire && haywire->is_exploding()) break;
+
+    auto bomb = dynamic_cast<Bomb*>(obj);
+    if (bomb) break;
+
+    auto goldbomb = dynamic_cast<GoldBomb*>(obj);
+    if (goldbomb && goldbomb->is_ticking()) break;
+
+    obj = nullptr;
+  }
+
+  if (!obj)
+  {
+    if (tstate == STATE_CORNERED)
+    {
+      set_action("recover", m_dir);
+      if (!m_sprite->animation_done()) return;
+    }
+    tstate = STATE_NORMAL;
+    m_physic.set_velocity_x(NORMAL_WALK_SPEED * (m_dir == Direction::LEFT ? -1 : 1));
+    m_physic.set_acceleration_x(0);
+    set_action(m_dir);
+    max_drop_height = NORMAL_MAX_DROP_HEIGHT;
+    set_walk_speed(NORMAL_WALK_SPEED);
+    return;
+  }
+
+  const Vector p1      = get_bbox().get_middle();
+  const Vector p2      = obj->get_bbox().get_middle();
+  const Vector vecdist = p2-p1;
+
+  if (glm::length(vecdist) > REALIZE_DIST && tstate == STATE_NORMAL) return;
+
+  switch (tstate)
+  {
+
+  case STATE_FLEEING:
+    if (m_dir == (vecdist.x > 0 ? Direction::LEFT : Direction::RIGHT)) return;
+    [[fallthrough]];
+
+  case STATE_NORMAL:
+  {
+    if (!on_ground()) break;
+
+    // Gold bomb is solid therefore raycast from
+    // one of the upper corners of the hitbox.
+    // (grown 1 just to make sure it doesnt interfere.)
+    const Rectf eye = get_bbox().grown(1.f);
+    if (!Sector::get().free_line_of_sight(
+      vecdist.x <= 0 ? eye.p1() : Vector(eye.get_right(), eye.get_top()),
+      obj->get_bbox().get_middle(),
+      false,
+      obj
+    )) break;
+
+    set_walk_speed(0);
+    m_physic.set_velocity_y(HOP_HEIGHT);
+    m_physic.set_velocity_x(0);
+    m_physic.set_acceleration_x(0);
+    m_dir = vecdist.x > 0 ? Direction::RIGHT : Direction::LEFT;
+    m_sprite->set_action("flee", m_dir);
+    tstate = STATE_REALIZING;
+    m_realize_timer.start(REALIZE_TIME);
+    break;
+  }
+
+  case STATE_REALIZING:
+    if (!m_realize_timer.check()) break;
+
+    flee(vecdist.x > 0 ? Direction::LEFT : Direction::RIGHT);
+    break;
+
+  default: break;
+
```

Comment:
The insides of the switch statement should be indented two spaces further. The unneeded empty lines at the beginning and end could also be removed.